From c5c2b98ba577cbb9235b985ed7c68125367b8783 Mon Sep 17 00:00:00 2001
From: Roland Haas <rhaas@tapir.caltech.edu>
Date: Thu, 31 Oct 2013 22:31:47 -0700
Subject: [PATCH 4/9] Hydro_Analysis: do reduction in global mode

also now works with different number of local components on each
processor
---
 schedule.ccl         | 22 ++++++++++++++++--
 src/Hydro_Analysis.c | 66 ++++++++++++++++++++++++++++------------------------
 2 files changed, 56 insertions(+), 32 deletions(-)

diff --git a/schedule.ccl b/schedule.ccl
index 688e3f2..079dab6 100644
--- a/schedule.ccl
+++ b/schedule.ccl
@@ -46,11 +46,29 @@ schedule Hydro_Analysis_Reduction IN Hydro_Analysis AFTER Hydro_Analysis_Prepare
 
 if (Hydro_Analysis_comp_rho_max || Hydro_Analysis_comp_rho_max_origin_distance)
 {
-  schedule Hydro_Analysis_LocationSearch IN Hydro_Analysis AFTER Hydro_Analysis_Reduction
+  schedule GROUP Hydro_Analysis_LocationSearch IN Hydro_Analysis AFTER Hydro_Analysis_Reduction
+  {
+  } "Look for the location of the maximum density"
+  schedule Hydro_Analysis_LocationSearch_Setup IN Hydro_Analysis_LocationSearch
+  {
+    LANG: C
+    OPTIONS: global
+  } "Prepare data structures for search"
+  # this is not perfect, loop-local calls us in all timelevels, not just
+  # the ones that a local routine would be called on, so we are mixing
+  # information from different times. Also not sure what happens if Carpet
+  # performs time interpolation for the maximum reduction. Fine is the maximum
+  # is on the finest level.
+  schedule Hydro_Analysis_LocationSearch_Search IN Hydro_Analysis_LocationSearch AFTER Hydro_Analysis_LocationSearch_Setup
   {
     LANG: C
     OPTIONS: global loop-local
-  } "Look for the location of the maximum density"
+  } "Search for the location of the maximum density"
+  schedule Hydro_Analysis_LocationSearch_Combine IN Hydro_Analysis_LocationSearch AFTER Hydro_Analysis_LocationSearch_Search
+  {
+    LANG: C
+    OPTIONS: global
+  } "Communicate and verify the location of the maximum density"
 }
 
 if (Hydro_Analysis_comp_rho_max_origin_distance)
diff --git a/src/Hydro_Analysis.c b/src/Hydro_Analysis.c
index 6512b84..c3ad2e8 100644
--- a/src/Hydro_Analysis.c
+++ b/src/Hydro_Analysis.c
@@ -83,23 +83,22 @@ void Hydro_Analysis_Reduction(CCTK_ARGUMENTS)
 }
 
 /* Look for the location of the density maximum */
-/* RH: this is scheduled global loop-local but does MPI calls. It will hang if
- * some processors have more components than others. */
-void Hydro_Analysis_LocationSearch(CCTK_ARGUMENTS)
+static CCTK_REAL local_rho_max_loc[4]; /* these two are shared between the next three routines */
+static int have_warned_about_multiple_maxima = 0;
+void Hydro_Analysis_LocationSearch_Setup(CCTK_ARGUMENTS)
 {
   DECLARE_CCTK_ARGUMENTS
   DECLARE_CCTK_PARAMETERS
 
-  static int have_warned_about_multiple_maxima = 0;
-
-  int set_maximum_location, warned_about_multiple_local_maxima = 0;
-
-  /* Initialized MPI-local quantities */
-  CCTK_REAL local_rho_max_loc[4], level_rho_max_loc[4];
   local_rho_max_loc[0] = 0.0;
   local_rho_max_loc[1] = 0.0;
   local_rho_max_loc[2] = 0.0;
-  local_rho_max_loc[3] = 0.0; // Information if it was found at all on this level
+  local_rho_max_loc[3] = 0.0; // Information if it was found at all
+}
+void Hydro_Analysis_LocationSearch_Search(CCTK_ARGUMENTS)
+{
+  DECLARE_CCTK_ARGUMENTS
+  DECLARE_CCTK_PARAMETERS
 
   /* Look for the location of the global maximum.
    * This algorithm will have problems when that occurs at more than
@@ -118,10 +117,10 @@ void Hydro_Analysis_LocationSearch(CCTK_ARGUMENTS)
         #pragma omp critical
         {
           // have to already warn here when we still have the actual coordinates around
-          if (verbosity_level >= 1 && round(local_rho_max_loc[3]) >= 1.)
+          if (verbosity_level >= 1 && local_rho_max_loc[3] >= 1.)
           { 
-            if(!have_warned_about_multiple_maxima && !warned_about_multiple_local_maxima)
-              CCTK_WARN(1, "Found more than one identical maximum in single patch.");
+            if(!have_warned_about_multiple_maxima && local_rho_max_loc[3] == 1.) // *second* time we get here
+              CCTK_WARN(1, "Found more than one identical maximum on single processor.");
             if (verbosity_level >= 2)
             {
               if (round(local_rho_max_loc[3]) == 1.) { // once we detect the second maximum, output the first as well
@@ -135,7 +134,6 @@ void Hydro_Analysis_LocationSearch(CCTK_ARGUMENTS)
                          x[i3D], y[i3D], z[i3D],
                          *Hydro_Analysis_rho_max);
             }
-            warned_about_multiple_local_maxima = 1;
           }
           local_rho_max_loc[0] += x[i3D];
           local_rho_max_loc[1] += y[i3D];
@@ -146,36 +144,44 @@ void Hydro_Analysis_LocationSearch(CCTK_ARGUMENTS)
     }
     LC_ENDLOOP3(location_reduction);
   }
-  /* After this at least one MPI process should have the location set.
+}
+void Hydro_Analysis_LocationSearch_Combine(CCTK_ARGUMENTS)
+{
+  DECLARE_CCTK_ARGUMENTS
+  DECLARE_CCTK_PARAMETERS
+
+  int set_maximum_location;
+
+  CCTK_REAL global_rho_max_loc[4];
+
+  /* At this point at least one MPI process should have the location set.
    * The others should have a value of 0.0, so that the sum reduction
    * gives the value on each MPI process. This however is not true sometimes
    * and this is the point where the algorithm can fail. */
-  /* RH: this assumption is false if a processor owns more than a single patch
-   * */
   CCTK_INT handle_sum = CCTK_LocalArrayReductionHandle("sum");
   if (handle_sum < 0)
      CCTK_WARN(0, "Unable to get reduction handle 'sum'.");
   if (CCTK_ReduceLocArrayToArray1D(cctkGH, -1, handle_sum, &local_rho_max_loc,
-                                   level_rho_max_loc, 4, CCTK_VARIABLE_REAL))
+                                   global_rho_max_loc, 4, CCTK_VARIABLE_REAL))
     CCTK_WARN(0, "Error while reducing local_rho_max_loc");
-  if (round(level_rho_max_loc[3]) > 0.) {
-    if (round(level_rho_max_loc[3]) == 1.)
+  if (round(global_rho_max_loc[3]) > 0.) {
+    if (round(global_rho_max_loc[3]) == 1.)
     {
       set_maximum_location = 1;
     } else {
       if (Hydro_Analysis_average_multiple_maxima_locations)
       {
-        level_rho_max_loc[0] /= level_rho_max_loc[3];
-        level_rho_max_loc[1] /= level_rho_max_loc[3];
-        level_rho_max_loc[2] /= level_rho_max_loc[3];
+        global_rho_max_loc[0] /= global_rho_max_loc[3];
+        global_rho_max_loc[1] /= global_rho_max_loc[3];
+        global_rho_max_loc[2] /= global_rho_max_loc[3];
 
         if (verbosity_level >= 1 && !have_warned_about_multiple_maxima)
         {
           CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
                      "Found more than one (%d) identical maximum, using average location (%g,%g,%g).",
-                     (int)lrint(level_rho_max_loc[3]),
-                     level_rho_max_loc[0], level_rho_max_loc[1],
-                     level_rho_max_loc[2]);
+                     (int)lrint(global_rho_max_loc[3]),
+                     global_rho_max_loc[0], level_rho_max_loc[1],
+                     global_rho_max_loc[2]);
           have_warned_about_multiple_maxima = 1;
         }
 
@@ -186,7 +192,7 @@ void Hydro_Analysis_LocationSearch(CCTK_ARGUMENTS)
           CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
                      "Found more than one (%d) identical maximum, not setting anything.",
                      (int)lrint(level_rho_max_loc[3]));
-          if (verbosity_level >= 2 && !warned_about_multiple_local_maxima)
+          if (verbosity_level >= 2 && local_rho_max_loc[3] == 1.0) // in this case we did not already warn about this locally
           { 
             CCTK_VWarn(1, __LINE__, __FILE__, CCTK_THORNSTRING,
                        "my candidate: (%g,%g,%g) with value %g.",
@@ -205,9 +211,9 @@ void Hydro_Analysis_LocationSearch(CCTK_ARGUMENTS)
 
   if (set_maximum_location)
   {
-      Hydro_Analysis_rho_max_loc[0] = level_rho_max_loc[0];
-      Hydro_Analysis_rho_max_loc[1] = level_rho_max_loc[1];
-      Hydro_Analysis_rho_max_loc[2] = level_rho_max_loc[2];
+      Hydro_Analysis_rho_max_loc[0] = global_rho_max_loc[0];
+      Hydro_Analysis_rho_max_loc[1] = global_rho_max_loc[1];
+      Hydro_Analysis_rho_max_loc[2] = global_rho_max_loc[2];
   }
 /*  CCTK_VInfo(CCTK_THORNSTRING, "New location: %g,%g,%g",
     Hydro_Analysis_rho_max_loc[0],
-- 
1.8.4.rc3

